src/main/scala/strawman/collection/Map.scala

```diff
@@ -0,0 +1,44 @@
+package strawman.collection
+
+import strawman.collection.mutable.Builder
+
+import scala.Option
+import scala.annotation.unchecked.uncheckedVariance
+import scala.Predef.???
+
+/** Base Map type */
+trait Map[K, +V]
+  extends Iterable[(K, V)]
```

Comment:
@dwijnand I agree, it's an interesting observation that the principle works both ways, and means opposite things depending on how you look at it. I think it would be good to find different names for each direction:

 - Principle of least power: Given some requirements, choose the most constraining ("obvious") API that fulfills the requirements, in order to maximize understanding what the implementation will do.

 - Principle of maximal abstraction: Given some requirements, choose the API that admits the largest set of possible implementations. 

Neither principle is inherently better than the other, but I believe that currently many libraries go too much towards maximal abstraction.